File: v2_fe/src/lib/resources.ts
/// The project's external resource document — mirrors the backend's
/// `taskflow-design/resources.rs` serde shapes exactly.
///
/// A *set* is a named, toggleable group of links ("Inter", "Analytics"): the
/// thing a user turns on or off. The links in it are the `<link>`/`<script>`
/// tags that set needs in the page's `<head>` — a web font's stylesheet plus
/// its two `preconnect`s, say. Toggling is what makes the feature useful: a font
/// can be kept in the document while contributing nothing to the page.
///
/// Tolerant read, strict write, exactly like `design-layout.ts`: the server
/// validates and self-heals, and this module only guarantees that a bad
/// response renders the empty document rather than throwing into the editor.

export type ResourceLink = {
  /// Present for `<link>` shapes. One of `preconnect`, `dns-prefetch`,
  /// `stylesheet` — there is no `preload`, because a preload without an `as`
  /// attribute fetches nothing and this model has no `as` field.
  rel?: string
  href?: string
  crossorigin: boolean
  /// Present for `<script>` shapes. A link carries ONE of `href`/`script`:
  /// the server refuses one carrying both, and an empty string counts as
  /// present, so the absent field must be ABSENT rather than blank.
  script?: string
  /// True when this link is a `<script src>` rather than a `<link>`.
  isScript: boolean
  isAsync: boolean
}

export type ResourceSet = {
  id: string
  name: string
  enabled: boolean
  links: ResourceLink[]
}

export type ResourcesDoc = {
  version: number
  sets: ResourceSet[]
}

export const DEFAULT_RESOURCES_DOC: ResourcesDoc = { version: 1, sets: [] }

/// The server's caps (`resources.rs`), mirrored so the editor cannot build a
/// document `validate` will refuse — the rule `createGroup` states for the
/// canvas layout. `MAX_HREF` is deliberately NOT here: see `appendLinks`.
export const MAX_SETS = 24
export const MAX_SET_NAME = 60
export const MAX_LINKS_PER_SET = 16

// ---------------------------------------------------------------------------
// Read-side parse
// ---------------------------------------------------------------------------

/// A non-blank string, or nothing. Blank counts as ABSENT, not as "": the
/// server treats a present-but-empty url field as present, which is how a blank
/// field turns into a refusal naming the wrong cause. This is a shape rule and
/// never a look at the url's own characters — scheme, length and NUL are the
/// server's to refuse, with a message the user can act on.
function str(v: unknown): string | undefined {
  return typeof v === "string" && v.trim() !== "" ? v : undefined
}

function normalizeLink(raw: unknown): ResourceLink | null {
  if (!raw || typeof raw !== "object") return null
  const l = raw as Record<string, unknown>
  const isScript = l.isScript === true
  const script = str(l.script)
  const href = str(l.href)
  const rel = str(l.rel)
  // A link survives on the address its OWN shape declares: `url_of` reads
  // exactly one of the two fields, so a href on a script shape would be
  // stranded somewhere no reader looks. Such a link (or one with no address at
  // all) cannot be stored by a validated write, and carrying it would only
  // resurface as "has no src" against a row the user cannot see.
  if (isScript ? !script : !href) return null
  const link: ResourceLink = {
    crossorigin: l.crossorigin === true,
    isScript,
    isAsync: l.isAsync === true,
  }
  if (isScript) {
    if (script) link.script = script
  } else {
    if (href) link.href = href
    if (rel) link.rel = rel
  }
  return link
}

function normalizeSet(raw: unknown): ResourceSet | null {
  if (!raw || typeof raw !== "object") return null
  const s = raw as Record<string, unknown>
  const id = str(s.id)
  const name = str(s.name)
  if (!id || !name) return null
  return {
    // Only an explicit `false` disables: absent is the server's own default
    // (`#[serde(default = "default_true")]`), and anything else is a document
    // this build should not read as "off".
    enabled: s.enabled !== false,
    id,
    name,
    links: (Array.isArray(s.links) ? s.links : []).flatMap((l): ResourceLink[] => {
      const link = normalizeLink(l)
      return link ? [link] : []
    }),
  }
}

/// Tolerant read of the stored document, for a response from a newer build or a
/// schema that moved under a cached tab. A client that renders the empty
/// document is recoverable; one that throws on load is not.
export function normalizeResources(raw: unknown): ResourcesDoc {
  if (!raw || typeof raw !== "object") return { version: 1, sets: [] }
  const obj = raw as Record<string, unknown>
  return {
    version:
      typeof obj.version === "number" && Number.isFinite(obj.version) ? obj.version : 1,
    sets: (Array.isArray(obj.sets) ? obj.sets : []).flatMap((s): ResourceSet[] => {
      const set = normalizeSet(s)
      return set ? [set] : []
    }),
  }
}

// ---------------------------------------------------------------------------
// Set edits — non-mutating, and identity means "nothing happened"
// ---------------------------------------------------------------------------

/// Set ids are opaque to the server; a counter plus a nonce is enough (they
/// only need to be unique inside one document and stable across a save).
/// Mirrors `design-layout.ts`'s `nextGroupId` exactly — the server refuses a
/// duplicate id by design, so collisions must not be possible here.
let setSeq = 0
function nextSetId(): string {
  setSeq += 1
  return `s${Date.now().toString(36)}${setSeq.toString(36)}`
}

/// Flip one set's toggle. An unknown id returns the document itself, so the
/// caller can tell "nothing happened" by identity.
export function toggleSet(doc: ResourcesDoc, id: string): ResourcesDoc {
  if (!doc.sets.some((s) => s.id === id)) return doc
  return {
    ...doc,
    sets: doc.sets.map((s) => (s.id === id ? { ...s, enabled: !s.enabled } : s)),
  }
}

/// Why `addSet` refuses this name, phrased for the person who typed it — or
/// `null` when it would accept. The rules live here, beside the checks they
/// explain, so the message and the refusal cannot drift apart, and the editor
/// can show the reason instead of an "Add set" that does nothing.
///
/// The server's own wording for these three is close (see `resources.rs`);
/// these are the UI's, and are not used as the save-time verdict.
export function setNameProblem(doc: ResourcesDoc, name: string): string | null {
  const trimmed = name.trim()
  if (!trimmed) return "A resource set needs a name."
  if ([...trimmed].length > MAX_SET_NAME) {
    return `A set name is limited to ${MAX_SET_NAME} characters.`
  }
  if (doc.sets.some((s) => s.name.trim().toLowerCase() === trimmed.toLowerCase())) {
    return `"${trimmed}" is already a set name.`
  }
  if (doc.sets.length >= MAX_SETS) return `At most ${MAX_SETS} resource sets.`
  return null
}

/// A new empty (enabled) set, or the document unchanged plus `id: ""` when the
/// name is refused — matching the server's rule, so the UI cannot build a
/// document the server will reject.
///
/// The name is measured in code points (`chars()` on the Rust side), not UTF-16
/// units: counting units would refuse a 60-emoji name the server accepts, and
/// that refusal would have no message anywhere.
export function addSet(doc: ResourcesDoc, name: string): { doc: ResourcesDoc; id: string } {
  if (setNameProblem(doc, name) !== null) return { doc, id: "" }
  const id = nextSetId()
  return {
    doc: {
      ...doc,
      sets: [...doc.sets, { id, name: name.trim(), enabled: true, links: [] }],
    },
    id,
  }
}

/// Remove a set. Its links go with it — unlike the canvas layout, there is no
/// ungrouped tail for them to fall back into.
export function removeSet(doc: ResourcesDoc, id: string): ResourcesDoc {
  const sets = doc.sets.filter((s) => s.id !== id)
  return sets.length === doc.sets.length ? doc : { ...doc, sets }
}

/// Remove one link from a set, addressed by position — the row the user is
/// looking at. Needed because `validate` refuses the ENTIRE document over a
/// single bad link and the manifest then contributes nothing: without this, one
/// stale `javascript:` url leaves every font in the project dead and the only
/// fix is to delete the whole set.
export function removeLink(doc: ResourcesDoc, setId: string, index: number): ResourcesDoc {
  const set = doc.sets.find((s) => s.id === setId)
  if (!set || index < 0 || index >= set.links.length) return doc
  return {
    ...doc,
    sets: doc.sets.map((s) =>
      s.id === setId ? { ...s, links: s.links.filter((_, i) => i !== index) } : s
    ),
  }
}

/// Append links to a set, stopping at `MAX_LINKS_PER_SET` and reporting what it
/// could not take. The document comes back by identity when nothing fit, so the
/// caller can distinguish "appended nothing" from "appended an empty list".
///
/// This is the ONLY path pasted links may take into a document. It caps the
/// set's length — an entity the UI creates — and deliberately does not inspect
/// a link's content: `MAX_HREF` is not mirrored, because "a url is limited to
/// 2048 characters" is a message the user can act on, and dropping what they
/// pasted would be a silence they cannot.
export function appendLinks(
  doc: ResourcesDoc,
  setId: string,
  links: ResourceLink[]
): { doc: ResourcesDoc; added: number; skipped: number } {
  const set = doc.sets.find((s) => s.id === setId)
  const room = set ? Math.max(0, MAX_LINKS_PER_SET - set.links.length) : 0
  const take = links.slice(0, room)
  if (!set || take.length === 0) return { doc, added: 0, skipped: links.length }
  return {
    doc: {
      ...doc,
      sets: doc.sets.map((s) => (s.id === setId ? { ...s, links: [...s.links, ...take] } : s)),
    },
    added: take.length,
    skipped: links.length - take.length,
  }
}

// ---------------------------------------------------------------------------
// Paste import
// ---------------------------------------------------------------------------

/// Tag bodies are matched with `[^>]*`: a `>` inside a quoted attribute value
/// would end the tag early, which is not worth a real tokenizer for a paste box
/// whose output the server validates anyway.
const TAG_RE = /<(link|script)\b([^>]*)>/gi

/// The standard HTML attribute pattern: quoted (either kind) or unquoted
/// values, plus bare attributes (`crossorigin`, `async`) parsed as `true`.
const ATTR_RE = /([a-zA-Z_:][-a-zA-Z0-9_:.]*)(?:\s*=\s*(?:"([^"]*)"|'([^']*)'|([^\s"'=<>`]+)))?/g

function parseAttrs(source: string): Record<string, string | true> {
  const attrs: Record<string, string | true> = {}
  ATTR_RE.lastIndex = 0
  let m: RegExpExecArray | null
  while ((m = ATTR_RE.exec(source))) {
    const value = m[2] ?? m[3] ?? m[4]
    attrs[m[1].toLowerCase()] = value === undefined ? true : value
  }
  return attrs
}

/// Read `<link>`/`<script>` tags out of pasted markup — the way a user actually
/// gets a Google Fonts set into this document: the font site hands them three
/// tags and they have nowhere to put them.
///
/// It PARSES and does not judge. A `javascript:` url comes through untouched,
/// because the server refuses it with a message the user can act on and a
/// client that silently dropped it would leave a dead box. Correspondingly it
/// applies no caps: `appendLinks` is the capped path, and the two concerns stay
/// separately testable.
///
/// The only thing "dropped" is a tag with no address at all (an inline
/// `<script>`, a `<link>` with no `href`): there is nowhere in the document
/// shape to put inline code, and a link with no url is refused server-side
/// against a row the user cannot see.
export function parsePastedLinks(text: string): ResourceLink[] {
  const links: ResourceLink[] = []
  if (!text) return links
  TAG_RE.lastIndex = 0
  let m: RegExpExecArray | null
  while ((m = TAG_RE.exec(text))) {
    const attrs = parseAttrs(m[2])
    const isScript = m[1].toLowerCase() === "script"
    const raw = attrs[isScript ? "src" : "href"]
    const url = typeof raw === "string" ? raw.trim() : ""
    if (!url) continue
    if (isScript) {
      // No `href`, no `rel`: see `ResourceLink.script`.
      links.push({
        crossorigin: attrs.crossorigin !== undefined,
        script: url,
        isScript: true,
        isAsync: attrs.async !== undefined,
      })
      continue
    }
    const link: ResourceLink = {
      crossorigin: attrs.crossorigin !== undefined,
      isScript: false,
      isAsync: false,
    }
    link.href = url
    const rel = typeof attrs.rel === "string" ? attrs.rel.trim() : ""
    // A missing `rel` is left out rather than defaulted to `stylesheet`: the
    // server's own refusal names the allowed set, which is a better hint than a
    // guess this module made for the user.
    if (rel) link.rel = rel
    links.push(link)
  }
  return links
}

File: backend/plugins/taskflow-design/src/resources.rs
//! The project's external resource document: named, toggleable groups of links
//! a page needs in order to look right — a web font and its companion
//! preconnects, or a third-party script.
//!
//! Stored as a `DesignFile` row at `styles/resources.json`, which means it
//! inherits versioning, the write validator, the operator and agent write
//! endpoints, and the manifest without any new plumbing. Shape mirrors
//! `tokens.rs`, which set that precedent.
//!
//! The security posture is the point of `validate`. These links are injected
//! into a document that runs scripts, so a URL scheme is a capability: only
//! `https:` is accepted, and `javascript:`/`data:` are refused in every
//! spelling. The sandbox is a separate origin with no cookies and a short-lived
//! read-only token, which bounds the blast radius — it does not license letting
//! an arbitrary scheme through.

use serde::{Deserialize, Serialize};

pub const RESOURCES_PATH: &str = "styles/resources.json";
pub const MAX_SETS: usize = 24;
pub const MAX_LINKS_PER_SET: usize = 16;
pub const MAX_HREF: usize = 2_048;
pub const MAX_SET_NAME: usize = 60;

/// `rel` values we accept on a `<link>`. All three are inert: none executes or
/// mutates the document, which is what makes them safe to allow alongside a
/// stylesheet. Every entry must also be USEFUL ON ITS OWN: `preload` is not,
/// because without an `as` attribute it fetches nothing and `ResourceLink` has
/// no `as` field, so it produced a `<link>` that sat in the DOM doing nothing.
/// Re-adding it means adding `as` to the model, not just the name back here.
pub const ALLOWED_REL: &[&str] = &["preconnect", "dns-prefetch", "stylesheet"];

/// camelCase on the wire throughout, so the frontend mirror is mechanical.
/// Everything is always serialised — no `skip_serializing_if` — because a
/// `None`/`false` round-trips through `#[serde(default)]` identically and the
/// attribute gymnastics buy nothing but a chance to get one wrong.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceLink {
    /// Present for `<link>` shapes.
    #[serde(default)]
    pub rel: Option<String>,
    #[serde(default)]
    pub href: Option<String>,
    #[serde(default)]
    pub crossorigin: bool,
    /// Present for `<script>` shapes.
    #[serde(default)]
    pub script: Option<String>,
    /// True when this link is a `<script src>` rather than a `<link>`.
    #[serde(default)]
    pub is_script: bool,
    #[serde(default)]
    pub is_async: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceSet {
    pub id: String,
    pub name: String,
    #[serde(default = "default_true")]
    pub enabled: bool,
    #[serde(default)]
    pub links: Vec<ResourceLink>,
}

fn default_true() -> bool { true }

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourcesDoc {
    #[serde(default = "default_version")]
    pub version: u32,
    #[serde(default)]
    pub sets: Vec<ResourceSet>,
}

fn default_version() -> u32 { 1 }

pub fn parse(raw: &str) -> Result<ResourcesDoc, String> {
    serde_json::from_str::<ResourcesDoc>(raw).map_err(|e| format!("invalid resources document: {e}"))
}

pub fn to_json_string(doc: &ResourcesDoc) -> String {
    serde_json::to_string(doc).expect("ResourcesDoc serialises")
}

/// The URL a link actually points at, whichever shape it is.
fn url_of(link: &ResourceLink) -> Option<&str> {
    if link.is_script { link.script.as_deref() } else { link.href.as_deref() }
}

/// `https:` and nothing else. Checked on the LOWERCASED url with leading C0
/// controls and spaces stripped, so case and leading whitespace cannot smuggle
/// a scheme past it, and deliberately strict — a scheme-relative `//host/x` and
/// a bare path are both refused, because "it will resolve to https anyway" is
/// an assumption about the page's origin, not a property of the link.
fn is_safe_url(url: &str) -> bool {
    // NOT `str::trim`, which strips UNICODE whitespace (NBSP, U+3000, and
    // friends) while a URL parser strips only C0 controls and space. Trimming
    // more than the parser does is not extra strictness — it is a hole: a
    // leading NBSP survives into the emitted value, the browser finds no
    // scheme there, and resolves the whole thing as a RELATIVE url, which is
    // the bare-path case above under a different spelling. So the predicate
    // below is the parser's own strip, and the guarantee it buys is scoped to
    // the SCHEME: for anything it accepts the value begins literally with
    // `https://`, and no later stage can alter those first eight characters —
    // every rewrite between here and a request either happens past index 8 or
    // is a deletion that cannot reach back into them.
    //
    // Two stages do diverge from the URL parser, and nothing here claims
    // otherwise. The HTML tokenizer normalises newlines BEFORE tokenizing, so
    // an interior CR is already gone when the URL parser's own strip runs; and
    // it decodes character references in attribute values, so `&amp;` is
    // reinterpreted. Both are scheme-preserving, and `&` is escaped wherever
    // these values are emitted, so neither becomes a live difference.
    //
    // U+0000 is the one character that CAN change the head: the tokenizer
    // rewrites it to U+FFFD before the URL parser's own strip sees it, so this
    // predicate cannot account for it. `validate` therefore refuses any url
    // containing one (see the containment check below) rather than leaving it
    // to the strip.
    let u = url.trim_matches(|c: char| c <= ' ').to_ascii_lowercase();
    u.starts_with("https://") && !u.starts_with("https://javascript:")
}

pub fn validate(doc: ResourcesDoc) -> Result<ResourcesDoc, String> {
    if doc.sets.len() > MAX_SETS {
        return Err(format!("at most {MAX_SETS} resource sets are allowed"));
    }
    let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();
    let mut seen_ids: std::collections::HashSet<String> = std::collections::HashSet::new();
    let mut sets = Vec::with_capacity(doc.sets.len());

    for set in doc.sets {
        let name = set.name.trim().to_string();
        if name.is_empty() {
            return Err("a resource set needs a name".into());
        }
        if name.chars().count() > MAX_SET_NAME {
            return Err(format!("a set name is limited to {MAX_SET_NAME} characters"));
        }
        if !seen.insert(name.to_lowercase()) {
            return Err(format!("the set name \"{name}\" is already used"));
        }
        let id = set.id.trim().to_string();
        if id.is_empty() {
            return Err("a resource set needs a non-empty id".into());
        }
        // Names are the human label and are checked above; the id is the key
        // every consumer addresses a set by, so a duplicate leaves the document
        // with no stable address for either set.
        if !seen_ids.insert(id.clone()) {
            return Err(format!("the set id \"{id}\" is already used"));
        }
        if set.links.len() > MAX_LINKS_PER_SET {
            return Err(format!("at most {MAX_LINKS_PER_SET} links per set"));
        }

        let mut links = Vec::with_capacity(set.links.len());
        for link in set.links {
            // One shape, one url field. `url_of` reads exactly one of them, so
            // a link carrying both would strand the other somewhere no reader
            // looks — the ambiguous shape is refused rather than resolved, and
            // refused before any scheme check, so the verdict does not depend
            // on which of the two fields happens to be the dangerous one.
            if link.href.is_some() && link.script.is_some() {
                return Err(format!(
                    "a link in \"{name}\" has both an href and a src; a link is one shape or the other"
                ));
            }
            let url = url_of(&link).ok_or_else(|| {
                format!("a link in \"{name}\" has no {}",
                    if link.is_script { "src" } else { "href" })
            })?;
            if url.trim().is_empty() {
                return Err(format!("a link in \"{name}\" has an empty url"));
            }
            // U+0000 is the one character `is_safe_url`'s strip cannot account
            // for, because the value reaches a TOKENIZER before it reaches a
            // URL parser: the tokenizer rewrites NUL to U+FFFD, which is
            // neither a C0 control nor a space, so it is not stripped and the
            // value resolves as a relative url while still reading as https
            // here. Refused wherever it appears rather than un-stripped, since
            // an interior NUL mangles the address just as silently.
            if url.contains('\u{0}') {
                return Err(format!(
                    "a url in \"{name}\" contains a U+0000 character; the html parser rewrites it to U+FFFD, so the address that loads would not be the one written"
                ));
            }
            if url.chars().count() > MAX_HREF {
                return Err(format!("a url in \"{name}\" is too long"));
            }
            if !is_safe_url(url) {
                return Err(format!("\"{url}\" is not an https address"));
            }
            if !link.is_script {
                let rel = link.rel.as_deref().unwrap_or("").trim().to_ascii_lowercase();
                if !ALLOWED_REL.contains(&rel.as_str()) {
                    // The alternatives are enumerated FROM the constant, not
                    // spelled out, so this message cannot drift from the list
                    // it describes — a refusal a user cannot act on is the
                    // thing `preload`'s removal was about.
                    return Err(format!(
                        "\"{rel}\" is not an allowed link relation; use one of: {}",
                        ALLOWED_REL.join(", ")
                    ));
                }
            }
            links.push(link);
        }
        sets.push(ResourceSet { id, name, enabled: set.enabled, links });
    }

    Ok(ResourcesDoc { version: doc.version, sets })
}

/// Every link from enabled sets, in document order, paired with whether it is a
/// script. Disabled sets contribute nothing — that is the whole point of the
/// toggle, and the reason a font can be kept without affecting a page.
pub fn enabled_links(doc: &ResourcesDoc) -> Vec<(bool, &ResourceLink)> {
    doc.sets
        .iter()
        .filter(|s| s.enabled)
        .flat_map(|s| s.links.iter().map(|l| (l.is_script, l)))
        .collect()
}
